src/cpp/dynamic-types/TypeObjectFactory.cpp

```diff
@@ -42,24 +43,51 @@ class TypeObjectFactoryReleaser
 
 };
 
+enum class TypeObjectFactoryInstanceState
+{
+    NOT_CREATED = 0,  // Instance has not been created
+    CREATING = 1,     // Instance is being created
+    CREATED = 2,      // Instance has been created
+    DESTROYING = 3    // Instance is being destroyed
+};
+
+static std::atomic<TypeObjectFactoryInstanceState> g_instance_state{TypeObjectFactoryInstanceState::NOT_CREATED};
```

Comment:
The problem is that, since destruction order is not guaranteed, the static recursive mutex could be destroyed before the `TypeObjectFactoryReleaser`, which could then lead to the mutex being used after destruction inside `TypeObjectFactory::delete_instance()`.

Atomic values have no destructor, and can be used during process exit procedure without issues.

Reply:
Good point. 
Did not know that about atomics, to be honest